### DISCUSSION TITLE
Recent update causes a hang on exit with a file open in a buffer

### DISCUSSION THREAD

**Discussion by duckpuppy** [LazyVim/LazyVim#4202] [Q&A]
After running an update in LazyVim yesterday (and I'm not sure how recently before that I had ran an update, but I tend to do it at least once a day) I started seeing a hang for 5-10 seconds when quitting neovim with a file buffer open. Exiting is instant without a buffer open. I've been trying to narrow down what's causing it, but was curious if others had seen anything like that? Even if I've had a buffer open as long as I destroy all open buffers before quitting it's instant, and destroying buffers doesn't hang.

EDIT: It seems to be any file, even one without any treesitter or LSP support running - I did briefly wonder if it was the LSP server hanging for a bit.

**Comment by dpetka2001** [comment]
Try to delete the temporary shada files in `~/.local/state/nvim/shada/` directory and see if that makes a difference. I don't observe any such problem you're describing.

**Comment by duckpuppy** [comment]
I've made another even stranger discovery: if I use the vim command bar - for any reason, even if it's a no-op like trying to move to the next buffer with only one buffer open - then I can exit immediately. If I don't use at least one vim command I see the hang. I tried editing a file and waiting a while to see if it's a timing issue and it doesn't seem to be, I still see the hang on exit if I don't enter a vim command... I'll try deleting the shada files and see if that helps.

EDIT: No change after deleting the shada files.

**Comment by duckpuppy** [comment]
I've now tried moving my existing ~/.config/nvim directory out of the way and using a fresh clone of `LazyVim/starter` and see the same behavior, so I can at least rule out something with my personal config.

**Comment by duckpuppy** [comment]
OK, now I'm really confused. I don't see any issues using `nvim --clean <file>` to skip all startup scripts, but I went as far as rolling back my starter checkout to the initial commit of the repo and still see the hang on exit. This just started yesterday for me...

**Comment by duckpuppy** [comment]
I think I've nailed it down - it happens in any git repo. I've been testing with various source code folders (easiest source of various types of files that would trigger different combinations of LSP clients and treesitter parsers) and my config folder, all of which are managed by git. Moving the .git directory out of the way eliminates the hang...

It does seem to be tied to LazyVim somehow... I've subbed in NvChad temporarily and the problem goes away.

**Comment by dpetka2001** [comment]
Try the following
```bash
1. mkdir ~/.config/lazytest
2. git clone https://github.com/LazyVim/starter ~/.config/lazytest
3. NVIM_APPNAME=lazytest nvim and wait for the installation to complete
4. While Lazy UI is still open press `C` to check and `U` to make sure everything is updated
5. Restart Neovim with command from step 3 
```
Do you still observe the same behavior? 

For cleanup
```bash
rm -rf ~/.config/lazytest
rm -rf ~/.local/share/lazytest
rm -rf ~/.local/state/lazytest
rm -rf ~/.cache/lazytest
```

**Reply by duckpuppy** [reply]
> ```shell
> NVIM_APPNAME=lazytest nvim
> ```

Yes, I still see the same behavior.

`NVIM_APPNAME=lazytest nvim <file>` on several files all show the same issue in git repos.

**Comment by duckpuppy** [comment]
It's almost definitely hanging in some shutdown code somewhere... the neovim display partially clears (all buffer text is gone, but the cursor, line number of the line the cursor is on, and status bar are still there) during the hang.

MANY EDITS: Man, I'm bad at typing today.

**Comment by duckpuppy** [comment]
I used the same methodology (changing the NVIM_APPNAME and .config location name for each) to try out a couple of other out-of-the-box vim configs, and I only see this behavior with LazyVim, which is sad because I like LazyVim the best.

**Reply by dpetka2001** [reply]
Well I don't see this behavior you're describing at all. I'm not able to reproduce this and not sure what causes your problem. But without being able to reproduce this, I'm out of ideas unfortunately.

**Comment by duckpuppy** [comment]
I left out a particularly relevant detail - I'm on macOS. I can repro this on two macOS machines, but I CANNOT repo this on either a Windows machine under WSL2/Debian or an actual Debian machine.

**Reply by dpetka2001** [reply]
Do you have the `yanky` Extra enabled?

**Reply by duckpuppy** [reply]
I do in my normal config, but obviously it's not in the `lazytest` config.

**Reply by dpetka2001** [reply]
What were the steps you did with the `lazytest` config?

**Reply by dpetka2001** [reply]
Also what Neovim version are you using?

**Reply by duckpuppy** [reply]
The steps were exactly what you said to do above.

```
NVIM v0.10.1
Build type: Release
LuaJIT 2.1.1713773202
```

**Reply by duckpuppy** [reply]
I do have a bit of confusion around the test, though - I cleaned up the test (using the 4 rm commands), re-cloned the repo into `lazytest` and used a different editor to change `~/.config/lazytest/lua/config/lazy.lua` so that lazy setup was just `require("lazy").setup({})` and ran `NVIM_APPNAME=lazytest nvim` - I thought that would stop LazyVim from initializing at all, but it did not. LazyVim still ran and all the included plugins were initialized as normal for a first-time startup (including the LazyVim Welcome notice).

**Reply by duckpuppy** [reply]
I'm going to clone LazyVim when I have more time and start trying to narrow down if there's a particular plugin that's causing it...

**Reply by dpetka2001** [reply]
Please do so, because I don't have MacOS myself to test it out and can't be of further help debugging this.

**Comment by duckpuppy** [comment]
I have some new findings - using my normal config, I started commenting out plugins in `~/.local/share/nvim/lazy/LazyVim/lua/lazyvim/plugins/*.lua` and if I comment out `noice`, the problem goes away completely. That also explains why it's limited to LazyVim.

**Reply by duckpuppy** [reply]
I think there's some correlation as to why entering the command buffer "fixes" things, because noice is invoked to display the command dialog. I don't know WHAT change that makes, though.

**Reply by duckpuppy** [reply]
For the moment, I'm going to leave that in place so the hang doesn't cause any extra friction... but of course I find this the day after @folke takes an extended vacation. Not that I begrudge that, it's just my luck. :) If I have time, I'll try to see why this problem started occurring.

**Reply by duckpuppy** [reply]
Just as a final follow-up, I ran `rm -Rf ~/.local/share/nvim/lazy/LazyVim` to allow lazy to re-clone LazyVim and undo all of my hacking and added `{ "folke/noice.nvim", enabled = false }` to my `.../plugins/init.lua` and the problem is gone. LazyVim isn't as nice without the message pop-ups, but it's usable until the problem is found and fixed.

**Reply by dpetka2001** [reply]
Yeah you'll just have to wait until Folke is back and maybe then create an issue at `noice` repo, so that he can have a look at it. What terminal are you using by the way? If it's `iterm`, have you tried other terminals like `kitty` or `wezterm` to see if it's some kind related to the terminal in use?

**Reply by duckpuppy** [reply]
I actually am using WezTerm, tried Kitty and basic Terminal. I actually don't have iTerm installed, but I don't THINK the terminal makes a difference (it didn't with the 3 I tried). I also tried different shells (bash, zsh, and fish).

**Reply by duckpuppy** [reply]
I like keeping things like terminals uniform across all my platforms, and wezterm is nice.

**Reply by duckpuppy** [reply]
[This](https://github.com/folke/noice.nvim/commit/1698725a663aca56bcd07a0e405bc441a5f6613b) commit seems highly suspicious.

**Reply by duckpuppy** [reply]
Yep. If I pin noice to the commit just prior to that one, no issues.

**Reply by dpetka2001** [reply]
Then you should create an issue with your findings at `noice` repo and be patient until Folke comes back and hopefully fixes it. It must be something specific to Mac, since on Linux Mint 21.3 I don't observe any abnormal behavior like this.

**Reply by duckpuppy** [reply]
https://github.com/folke/noice.nvim/issues/921

**Reply by ArchiFleKs** [reply]
Thanks I actually have the same issue at the moment on mac os 14.5 and nvim 0.10.1

**Reply by auvn** [reply]
Hello guys! do you have this issue resolved? Same problem on my end

**Reply by dpetka2001** [reply]
To my understanding this was an issue related to `noice.nvim` mostly specific in MacOS. I could not reproduce at the time on my Linux Mint. And right now I don't observe any kind of behavior related to this. 

Try disabling `noice.nvim` temporarily to see if your problem goes away. Also, try deleting your shada files.

**Reply by frogiraffe** [reply]
> To my understanding this was an issue related to `noice.`

yes this helped a lot, im on arch linux tho
